src/canard.c

```diff
@@ -1,6 +1,845 @@
+    }
+  } else
+  {
+    can_id = ((uint32_t)priority << 24) | ((uint32_t)data_type_id << 8) | (uint32_t)canardGetLocalNodeID(ins);
+    
+    if (payload_len > 7)
+    {
+      crc = crc_add_signature(crc, data_type_signature);
+      crc = crc_add(crc, payload, payload_len);
+    }
+  }
+
+  canardEnqueueData(ins, can_id, inout_transfer_id, crc, payload, payload_len);
+
+  tidIncrement(inout_transfer_id);
+
+  return 1;
+}
+
+/**
+ * Sends a request or a response transfer.
+ * Fails if the node is in passive mode.
+ */
+int canardRequestOrRespond(CanardInstance* ins,
+                            uint8_t destination_node_id,
+                            uint64_t data_type_signature,
+                            uint16_t data_type_id,
+                            uint8_t* inout_transfer_id,
+                            uint8_t priority,
+                            CanardRequestResponse kind,
+                            const void* payload,
+                            uint16_t payload_len)
+{
+  if (payload == NULL)
+  {
+      return -1;
+  }
+  if (canardGetLocalNodeID(ins) == 0)
+  {
+    return -1;
+  }
+  if (priority > 31)
+  {
+      return -1;
+  }
+
+  const uint32_t can_id = ((uint32_t)priority << 24) | ((uint32_t)data_type_id << 16) | 
+                            ((uint32_t)kind << 15) | ((uint32_t)destination_node_id << 8) | 
+                              (1 << 7) | (uint32_t)canardGetLocalNodeID(ins);
+  uint16_t crc = 0xFFFFU;
+
+  if (payload_len > 7)
+  {
+    crc = crc_add_signature(crc, data_type_signature);
+    crc = crc_add(crc, payload, payload_len);
+  }
+
+  canardEnqueueData(ins, can_id, inout_transfer_id, crc, payload, payload_len);
+
+  tidIncrement(inout_transfer_id);
+
+  return 1;
+}
+
+/**
+ * Returns a pointer to the top priority frame in the TX queue.
+ * Returns NULL if the TX queue is empty.
+ */
+const CanardCANFrame* canardPeekTxQueue(const CanardInstance* ins)
+{
+  if (ins->tx_queue == NULL)
+  {
+    return NULL;
+  }
+  return &ins->tx_queue->frame;
+}
+
+/**
+ * Removes the top priority frame from the TX queue.
+ */
+void canardPopTxQueue(CanardInstance* ins) 
+{
+  CanardTxQueueItem* item = ins->tx_queue;
+  ins->tx_queue = item->next;
+  canardFreeBlock(&ins->allocator, item);
+}
+
+/**
+ * Processes a received CAN frame with a timestamp.
+ */
+void canardHandleRxFrame(CanardInstance* ins, const CanardCANFrame* frame, uint64_t timestamp_usec)
+{
+  
+  uint8_t transfer_type = canardTransferType(frame->id);
+  uint8_t priority = CANARD_PRIORITY_FROM_ID(frame->id);
+  uint8_t source_node_id = CANARD_SOURCE_ID_FROM_ID(frame->id);
+  uint8_t destination_node_id = (transfer_type == CanardTransferTypeBroadcast) ? 0 : CANARD_DEST_ID_FROM_ID(frame->id);
+  uint16_t data_type_id = canardDataType(frame->id);
+  uint32_t transfer_descriptor = CANARD_MAKE_TRANSFER_DESCRIPTOR(data_type_id, transfer_type, source_node_id, destination_node_id);
+  
+  CanardRxState* rxstate;
```

Comment:
This variable leaks past the `if-else` below uninitialized. MISRA and other high-integrity conding standards require to default-initialize all variables to prevent that.

The line should be `CanardRxState* rxstate = NULL;`